include/ppx/knob.h

```diff
@@ -0,0 +1,377 @@
+// Copyright 2023 Google LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+#ifndef ppx_knob_h
+#define ppx_knob_h
+
+#include "ppx/command_line_parser.h"
+#include "ppx/config.h"
+#include "ppx/imgui_impl.h"
+#include "ppx/log.h"
+
+#include <iostream>
+#include <memory>
+#include <unordered_set>
+#include <vector>
+
+namespace ppx {
+
+// ---------------------------------------------------------------------------------------------
+// Knob Classes
+// ---------------------------------------------------------------------------------------------
+
+// Knob is an abstract class which contains common features for all knobs and knob hierarchy
+class Knob
+{
+public:
+    Knob(const std::string& flagName)
+        : mFlagName(flagName), mDisplayName(flagName), mIndent(0) {}
+    virtual ~Knob() = default;
+
+    const std::string& GetDisplayName() const { return mDisplayName; }
+    const std::string& GetFlagName() const { return mFlagName; }
+    const std::string& GetFlagHelp() const { return mFlagDesc; }
+    size_t             GetIndent() const { return mIndent; }
+
+    void SetDisplayName(const std::string& displayName) { mDisplayName = displayName; }
+    void SetFlagDesc(const std::string& flagDesc) { mFlagDesc = flagDesc; }
+    void SetIndent(size_t indent) { mIndent = indent; }
+
+    virtual void Draw() = 0;
+
+    // Returns the flag description used in usage message
+    virtual std::string GetFlagHelpText() const = 0;
+
+    // Updates knob value from commandline flag
+    virtual void UpdateFromFlags(const CliOptions& opts) = 0;
+
+    virtual void ResetToDefault() = 0;
+
+protected:
```

Comment:
There's no point on making those protected, you already have public accessor functions.